apis/oam/v1alpha2/types.go

```diff
@@ -0,0 +1,857 @@
+	// all workload kinds.
+	// +optional
+	AppliesToWorkloads []string `json:"appliesToWorkloads,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// A TraitDefinition registers a kind of Kubernetes custom resource as a valid
+// OAM trait kind by referencing its CustomResourceDefinition. The CRD is used
+// to validate the schema of the trait when it is embedded in an OAM
+// ApplicationConfiguration.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type TraitDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec TraitDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// TraitDefinitionList contains a list of TraitDefinition.
+type TraitDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []TraitDefinition `json:"items"`
+}
+
+// A ScopeDefinitionSpec defines the desired state of a ScopeDefinition.
+type ScopeDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this scope kind.
+	Reference DefinitionReference `json:"definitionRef"`
+
+	// AllowComponentOverlap specifies whether an OAM component may exist in
+	// multiple instances of this kind of scope.
+	AllowComponentOverlap bool `json:"allowComponentOverlap"`
+}
+
+// +kubebuilder:object:root=true
+
+// A ScopeDefinition registers a kind of Kubernetes custom resource as a valid
+// OAM scope kind by referencing its CustomResourceDefinition. The CRD is used
+// to validate the schema of the scope when it is embedded in an OAM
+// ApplicationConfiguration.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type ScopeDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ScopeDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ScopeDefinitionList contains a list of ScopeDefinition.
+type ScopeDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ScopeDefinition `json:"items"`
+}
+
+// A ComponentParameter defines a configurable parameter of a component.
+type ComponentParameter struct {
+	// Name of this parameter. OAM ApplicationConfigurations will specify
+	// parameter values using this name.
+	Name string `json:"name"`
+
+	// FieldPaths specifies an array of fields within this Component's workload
+	// that will be overwritten by the value of this parameter. The type of the
+	// parameter (e.g. int, string) is inferred from the type of these fields;
+	// All fields must be of the same type. Fields are specified as JSON field
+	// paths without a leading dot, for example 'spec.replicas'.
+	FieldPaths []string `json:"fieldPaths"`
+
+	// TODO(negz): Use +kubebuilder:default marker to default Required to false
+	// once we're generating v1 CRDs.
```

Comment:
This could also be changed from a `*bool` to a `bool` if we wanted default to `false`. However, that would be a divergence from the Kubernetes [api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) so I would support manually handling this until default tag is available as you suggest :+1: